bin/picom-trans

```diff
@@ -46,194 +55,311 @@ print_usage() {
   exit 0
+} #}}
+
+delete_opacity()
+{ #{{
+  xprop -id "$winid" -remove _NET_WM_WINDOW_OPACITY
+  exit 0
+} #}}
+
+reset_opacity()  # Reset opacity of all windows
+{ #{{
+  for wid in $(xwininfo -root -tree | sed -n 's/^     \(0x[[:xdigit:]]*\).*/\1/p')
+  do xprop -id "$wid" -remove _NET_WM_WINDOW_OPACITY 2>/dev/null
+  done
+  exit 0
+} #}}
+
+set_opacity()
+{ #{{
+  if ! echo "$target_opacity" | grep -qE '^[+-]?[[:digit:]]+%?$'
+  then
+    if test -z "$target_opacity"
+    then echo "No opacity specified" >&2
+    else echo "Invalid opacity specified: $target_opacity" >&2
+    fi
+    exit 1
+  fi
+
+  # strip trailing '%' sign, if any
+  target_opacity="${target_opacity%%%}"
+
+  current_opacity=$(get_current_opacity)
+  echo "$target_opacity" | grep -qE '^[+-]' \
+    && target_opacity=$(( current_opacity + target_opacity ))
+
+  test $target_opacity -lt 0   && target_opacity=0
+  test $target_opacity -gt 100 && target_opacity=100
+
+  target_opacity=$(( target_opacity * 0xffffffff / 100 ))
+  xprop -id "$winid" -f _NET_WM_WINDOW_OPACITY 32c \
+      -set _NET_WM_WINDOW_OPACITY "$target_opacity"
+
+  exit $?
+} #}}
+
+toggle_opacity()
+{ #{{
+  current_opacity="$(xprop -id "$winid" -notype _NET_WM_WINDOW_OPACITY \
+    | sed -E 's/^_NET_WM_WINDOW_OPACITY = ([0-9]*)$|^.*$/\1/')"
+
+  if test -z "$current_opacity"
+  then
+    test -z "$target_opacity" && target_opacity=100
+    set_opacity
+  else
+    delete_opacity
+  fi
+} #}}
+
+
+# Warn about rename of compton to picom
+case "$0" in
+  *compton-trans*)  echo "Warning: compton has been renamed, please use picom-trans instead" >&2;;
+esac
+
+
+# "command" is a shell built-in, faster than "which"
+if test -z "$(command -v xprop)" -o -z "$(command -v xwininfo)"; then
+  echo "The command xwininfo or xprop is not available. They might reside in a package named xwininfo, xprop, x11-utils, xorg-xprop, or xorg-xwininfo." >&2
+  exit 1
+fi
+
+
+# No arguments given. Show help.
+if test $# -eq 0; then
+  print_usage >&2
+  exit 1
 fi
 
-# Calculate the desired opacity
-if echo "$opacity" | grep -q '^[+-]'; then
-  opacity=$((cur + opacity))
+
+# Variables
+#   action is set to 'set' by default
+action=set
+winid=
+winidtype=
+target_opacity=
+
+# If there's only one argument, and it's a valid opacity
+# then take it as target_opacity.  Else, parse all arguments.
+if test $# -eq 1 && echo "$1" | grep -qE '^[+-]?[[:digit:]]+%?$'
+then
+  target_opacity=$1
+  shift
+else
+  parse_args "$@"
 fi
 
-test $opacity -lt 0 && opacity=0
-test $opacity -gt 100 && opacity=100
 
-# Set opacity
-opacity=$((opacity * 0xffffffff / 100))
-xprop -id "$topmost" -f _NET_WM_WINDOW_OPACITY 32c \
-  -set _NET_WM_WINDOW_OPACITY "$opacity"
+# reset_opacity doesn't need $winid
+case $action in
+  (reset)   reset_opacity;;
+esac
+
+# Any other action needs $winid
+# NOTE: Do NOT change the order of winid= and winidtype= below
+#       the output of get_target_window_id depends on $winidtype
+winid=$(get_target_window_id)
```

Comment:
`get_target_window_id` can fail (i.e. window with name not found), but calling `exit` in the sub-shell does not exit in the calling context.
```suggestion
winid=$(get_target_window_id) || exit 1
```

Reply:
Implemented in latest commit.